### ISSUE TITLE
Fail to generate schema when using tsconfig-paths

### ISSUE THREAD

**Issue by marcelo-agil** [YousefED/typescript-json-schema#505]
Hi guys,

I'm  using tsconfig-paths to avoid path hell. Instead of importing files using import * from '../../../', it uses an array of path aliases and they current values.  Works pretty well.

Example of tsconfig-path:
```
"paths": {
      "@sharedDomain/*": ["src/modules/shared/domain/*"],
}

```
The problem is when I try to programmatic use typescript-json-schema, it throws this error:
`"Cannot find module '@shared/util/util.interface' or its corresponding type declarations.".`

The file above constains this code:
```
import * as IUtil from '@shared/util/util.interface';

export interface IUserAgenda extends IUtil.IDateFields {
  /**
   * Inform if this user has agenda app configured
   *@default false
   */
  isAgendaAppIntegrated?: boolean;
  /**
   * App key
   *
   */
  agendaAppKey?: string;
  /**
   * App secret
   *
   */
  agendaAppSecret?: string;
  /**
   * App endpoint
   *
   */
  agendaAppEndpoint?: string;
}

```

My question is:
How to make typescript-json-schema understand the tsconfig-paths?

**Comment by erkin98** [comment]
try changing
```
"paths": {
      "@sharedDomain/*": ["src/modules/shared/domain/*"],
} 
```
to
```
"paths": {
      "@shared/*": ["src/modules/shared/*"],
}

**Comment by marcelo-agil** [comment]
@erki

> try changing
> 
> ```
> "paths": {
>       "@sharedDomain/*": ["src/modules/shared/domain/*"],
> } 
> ```
> 
> to
> 
> ```
> "paths": {
>       "@shared/*": ["src/modules/shared/*"],
> }
> ```

My mistake on the example. The code is right, but won't work.
 In the interface file:
` import * as IUtil from '@shared/util/util.interface';`

In tsconfig-paths:
_"@shared/*": ["src/modules/shared/*"],_

It doesn't recognize. But on runtime code works, file gets imported.
If I used full path, than typescript-json-schema works.

He is not translating @shared. How can I achieve that?

**Comment by erkin98** [comment]
could you specify which command are you running and would you consider pasting your tsconfig?

**Comment by marcelo-agil** [comment]
> could you specify which command are you running and would you consider pasting your tsconfig?

Sure.

Here's my tsconfig.ts:
```
{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "lib": ["es6"],
    "allowJs": true,
    "outDir": "./dist",
    "rootDir": "./src",
    "removeComments": true,
    "typeRoots": ["./node_modules/@types", "./src/@types"],
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "baseUrl": ".",
    "paths": {
      "@sharedDomain/*": ["src/modules/shared/domain/*"],
      "@negotiation/*": ["src/modules/user/*"],
      "@notification/*": ["src/modules/notification/*"],
      "@customer/*": ["src/modules/customer/*"],
      "@person/*": ["src/shared/person/*"],
      "@personDomain/*": ["src/modules/shared/person/domain/*"],
      "@personController/*": [
        "src/modules/shared/person/app/http/controller/*"
      ],
      "@personHelper/*": ["src/modules/user/app/http/helper/*"],
      "@security/*": ["src/modules/shared/security/*"],
      "@userDomain/*": ["src/modules/shared/security/domain/user/*"],
      "@product/*": ["src/modules/product/*"],
      "@scheduler/*": ["src/modules/scheduler/*"],
      "@shared/*": ["src/modules/shared/*"],
      "@config/*": ["src/config/*"]
    }
  },
  "include": ["src/**/*", "server.ts"]
}

```

The command is run through script:

```
const program = TJS.getProgramFromFiles(
      [
        resolve(
          `${@shared}/security/domain/user/user-agenda.interface.ts`
        ),
      ],
      compilerOptions
    );
    const schema = TJS.generateSchema(program, '*', compilerOptions);
```

The "Agenda" interface is just a test file:
```

import * as IUtil from '@shared/util/util.interface';

export interface IUserAgenda extends IUtil.IDateFields {
  /**
   * Inform if this user has agenda app configured
   *@default false
   */
  isAgendaAppIntegrated?: boolean;
  /**
   * App key
   *
   */
  agendaAppKey?: string;
  /**
   * App secret
   *
   */
  agendaAppSecret?: string;
  /**
   * App endpoint
   *
   */
  agendaAppEndpoint?: string;
}

```

Once again, if I replace the @shared by the path, it works.

**Comment by marcelo-agil** [comment]
@erkin98 any thoughts?

**Comment by erkin98** [comment]
I suggest you use the `typescript-json-schema "tsconfig.json" "*"  -o all.json and add `"types": ["folder-inside-src/@types"],`

**Comment by nevadavid** [comment]
Any news on this issue? Because I use v0.62.0 but I get same error with similar context.

**Comment by DarkRoku12** [comment]
The thing is that when using `TJS.getProgramFromFiles` TypeScript compiler will actually include only definitions explicitly mentioned.

If you check: https://github.com/YousefED/typescript-json-schema/blob/43e557d9c50dc1695dfba1ff83fc50f40f9376ca/typescript-json-schema.ts#L1795-L1822

You'll see that `const configParseResult = ts.parseJsonConfigFileContent(...)` 
and then `configParseResult.fileNames` is basically resolving the glob pattern to an actual file list (all matching files).

Basically, the compiler expects to include all the files in your project.
An easy solution is to just copy bring that function into your own project source code and use the generated TS Program directly.

Then `tsconfig.paths` will work as intended.

**Comment by felipecsl** [comment]
> The thing is that when using `TJS.getProgramFromFiles` TypeScript compiler will actually include only definitions explicitly mentioned.
> 
> If you check:
> 
> [typescript-json-schema/typescript-json-schema.ts](https://github.com/YousefED/typescript-json-schema/blob/43e557d9c50dc1695dfba1ff83fc50f40f9376ca/typescript-json-schema.ts#L1795-L1822)
> 
> Lines 1795 to 1822 in [43e557d](/YousefED/typescript-json-schema/commit/43e557d9c50dc1695dfba1ff83fc50f40f9376ca)
> 
>  export function programFromConfig(configFileName: string, onlyIncludeFiles?: string[]): ts.Program { 
>      // basically a copy of https://github.com/Microsoft/TypeScript/blob/3663d400270ccae8b69cbeeded8ffdc8fa12d7ad/src/compiler/tsc.ts -> parseConfigFile 
>      const result = ts.parseConfigFileTextToJson(configFileName, ts.sys.readFile(configFileName)!); 
>      const configObject = result.config; 
>   
>      const configParseResult = ts.parseJsonConfigFileContent( 
>          configObject, 
>          ts.sys, 
>          path.dirname(configFileName), 
>          {}, 
>          path.basename(configFileName) 
>      ); 
>      const options = configParseResult.options; 
>      options.noEmit = true; 
>      delete options.out; 
>      delete options.outDir; 
>      delete options.outFile; 
>      delete options.declaration; 
>      delete options.declarationDir; 
>      delete options.declarationMap; 
>   
>      const program = ts.createProgram({ 
>          rootNames: onlyIncludeFiles || configParseResult.fileNames, 
>          options, 
>          projectReferences: configParseResult.projectReferences, 
>      }); 
>      return program; 
>  } 
> You'll see that `const configParseResult = ts.parseJsonConfigFileContent(...)` and then `configParseResult.fileNames` is basically resolving the glob pattern to an actual file list (all matching files).
> 
> Basically, the compiler expects to include all the files in your project. An easy solution is to just copy bring that function into your own project source code and use the generated TS Program directly.
> 
> Then `tsconfig.paths` will work as intended.

Nice, this worked. Full example below:
```typescript
function typescriptTypeToJsonSchema(
  typescriptType: string,
  typesFileAbsPath: string
): JsonSchema | null {
  const settings: PartialArgs = { required: true };
  // Gotcha: We need to use tsconfig.json from the root of the project, not the temporal package
  const tsconfigFilePath = 'tsconfig.json';
  const files = [typesFileAbsPath.replace('.js', '.ts')];
  try {
    const program = programFromConfig(path.join(process.cwd(), tsconfigFilePath), files);
    // @ts-expect-error - TODO: fix this
    return generateSchema(program, typescriptType, settings) as JsonSchema;
  } catch (e) {
    console.error(`Failed to generate JSON schema for ${typescriptType} at ${typesFileAbsPath}`, e);
    throw e;
  }
}

function programFromConfig(configFileName: string, onlyIncludeFiles?: string[]): ts.Program {
  // basically a copy of https://github.com/Microsoft/TypeScript/blob/3663d400270ccae8b69cbeeded8ffdc8fa12d7ad/src/compiler/tsc.ts -> parseConfigFile
  const result = ts.parseConfigFileTextToJson(configFileName, ts.sys.readFile(configFileName)!);
  const configObject = result.config;

  const configParseResult = ts.parseJsonConfigFileContent(
    configObject,
    ts.sys,
    path.dirname(configFileName),
    {},
    path.basename(configFileName)
  );
  const options = configParseResult.options;
  options.noEmit = true;
  delete options.out;
  delete options.outDir;
  delete options.outFile;
  delete options.declaration;
  delete options.declarationDir;
  delete options.declarationMap;

  const program = ts.createProgram({
    rootNames: onlyIncludeFiles || configParseResult.fileNames,
    options,
    projectReferences: configParseResult.projectReferences,
  });
  return program;
}

```
